src/ralph/virtual/management/commands/openstack_sync.py

```diff
@@ -0,0 +1,295 @@
+
+                for ip in (
+                    set(ralph_servers[server_id]['ips'])-set(
+                        servers[server_id]['ips'])
+                ):
+                    del_ip = IPAddress.objects.get(address=ip)
+                    del_ip.asset = None
+                    del_ip.save()
+                    modified = True
+
+                if modified:
+                    self.summary['mod_instances'] += 1
+
+            # ADD missing server
+            else:
+                server = servers[server_id]
+                new_server = CloudHost(
+                    hostname=server['hostname'],
+                    parent=project,
+                    host_id=server_id,
+                )
+                new_server.save()
+                self.summary['new_instances'] += 1
+                for ip in server['ips']:
+                    self._add_ip(server, ip, False)
+
+        # remove servers
+        try:
+            for server_id in (
+                self.ralph_projects[project.project_id]['servers'].keys(
+                ) - servers.keys()
+            ):
+                CloudHost.objects.get(host_id=server_id).delete()
+                self.summary['del_instances'] += 1
+        except KeyError:
+            pass
+
+    def _add_project(self, data, project_id):
+        """Add/modify project to openstack"""
+        if project_id in self.ralph_projects.keys():
+            project = CloudProject.objects.get(project_id=project_id)
+            if (
+                self.ralph_projects[project_id][
+                    'cloudprovider'] != self.cloud_provider or
+                self.ralph_projects[project_id]['name'] != data['name'] or
+                self.openstack_projects[project_id]['tag']
+                    not in self.ralph_projects[project_id]['tags']
+            ):
+                self.summary['mod_projects'] += 1
+                project.cloudprovider = self.cloud_provider
+                project.name = data['name']
+                project.save()
+                project.tags.add(data['tag'])
+
+        else:
+            self.summary['new_projects'] += 1
+            project = CloudProject(
+                name=data['name'],
+                project_id=project_id,
+                cloudprovider=self.cloud_provider,
+            )
+            project.save()
+            project.tags.add(data['tag'])
+        self._add_servers(data['servers'], project)
+
+    def _update_ralph(self):
+        """Update existing and add new ralph data"""
+
+        for project_id in self.openstack_projects:
+            self._add_project(
+                self.openstack_projects[project_id],
+                project_id,
+            )
+
+    def _cleanup(self):
+        """
+        Remove all projects that doesn't exist in openstack from ralph
+        """
+        for project_id in (set(self.ralph_projects.keys()) - set(
+            self.openstack_projects.keys())
+        ):
+            # TODO: Wrzucic do revison
```

Comment:
try use ``revision_context_manager.start()`` and ``revision_context_manager.end()`` from django-reversion (ex. https://github.com/etianen/django-reversion/blob/8c0113d1bb18cc9d06d7724df3102914877a3ff9/src/reversion/middleware.py#L22)